### PR TITLE
[codex] add OTLP and file IO contract oracles

### PR DESCRIPTION
**Pull request by strawgate**
## What changed

This PR upgrades the OTLP/file IO correctness work from a narrow loopback check into a contract-focused verification slice.

It adds:

- OTLP receiver contract tests for JSON, protobuf, and zstd semantic equivalence
- replay-style file boundary tests for newline checkpointing and per-source remainder isolation
- an official-protobuf oracle path for OTLP encoding/decoding semantics
- black-box `otelcol-contrib` oracle tests in both directions:
  - `our OtlpSink -> official otelcol receiver`
  - `official otelcol sender -> our OtlpReceiverInput`
- a written adapter contract and worker notes documenting the boundary guarantees and remaining gaps

It also fixes OTLP receiver semantic drift around `bytesValue` handling and unsupported protobuf attribute emission.

## Why

The goal is to stop treating OTLP and file IO correctness as a series of one-off regressions.

This change moves the repo toward executable boundary contracts:

- compare semantics instead of just happy-path transport success
- validate against official OTLP protobuf types and the official Collector, not only our own implementations
- make file checkpoint and partial-line invariants explicit and testable

## Impact

- better protection against OTLP protocol drift
- stronger confidence that file input behavior is stable across newline and multi-source edge cases
- a reusable external-oracle harness for future gzip/gRPC and replay-compatibility work

## Validation

- `cargo test -p logfwd-io --test it`
- `cargo test -p logfwd-io --test it otelcol_ -- --ignored --nocapture`
- `cargo test -p logfwd-output encode_batch_matches_official_proto_oracle -- --nocapture`
- `cargo test -p logfwd-io otlp_receiver_preserves_semantics_across_json_protobuf_and_zstd -- --nocapture`

## Remaining follow-up

- OTLP gzip parity
- OTLP gRPC parity
- external oracle coverage for file replay/copytruncate behavior

> [!NOTE]
> ### Add OTLP and file IO contract oracle tests for the `logfwd-io` crate
> - Adds integration tests validating that the OTLP receiver preserves semantics across JSON, protobuf, and zstd-compressed protobuf wire formats, and returns 400 for invalid payloads (e.g. bad `bytesValue`).
> - Adds a loopback contract test that sends an Arrow batch through `OtlpSink` and asserts field preservation after round-trip through `OtlpReceiverInput` with Zstd compression.
> - Adds file boundary contract tests verifying that `FramedInput` over `FileInput` advances checkpoints only at newline boundaries and isolates per-source state for globbed files.
> - Adds two ignored black-box oracle tests in [otelcol_blackbox_contract.rs](https://github.com/strawgate/memagent/pull/1290/files#diff-e9d061fbe715ea7e0995ad9de60f1524def9d8fabdf0e590d5bfa8210f45da9a) that download the official OpenTelemetry Collector binary and validate our sink/receiver against its behavior.
> - Fixes `decode_otlp_logs_json` to preserve numeric and boolean types in attributes, decode `bytesValue` from base64 to hex, skip unsupported value kinds, and return errors for invalid `intValue` or `bytesValue` instead of emitting malformed fields.
> - Behavioral Change: OTLP JSON requests with invalid `intValue` or `bytesValue` now return HTTP 400 and are not enqueued; previously they emitted empty or malformed fields.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 6f9de35.</sup>
> 
>

### CODE REVIEW

File: crates/logfwd-io/src/otlp_receiver.rs [hunk 1]
```diff
@@ -402,33 +405,87 @@ fn decode_otlp_logs_json(body: &[u8]) -> Result<Vec<u8>, InputError> {
...
+        out.push(b'"');
+        out.extend_from_slice(key.as_bytes());
+        out.extend_from_slice(b"\":");
+        if let Some(s) = i.as_str() {
+            out.extend_from_slice(s.as_bytes());
+        } else if let Some(n) = i.as_i64() {
+            write_i64_to_buf(out, n);
+        } else {
+            out.extend_from_slice(b"0");
+        }
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/logfwd-io/src/otlp_receiver.rs]: Fixed in `68e24be8`.

`intValue` with an unexpected JSON type now returns an `InputError` instead of silently emitting `0`, and I added a regression test for that malformed input path.

File: crates/logfwd-io/src/otlp_receiver.rs [hunk 2]
```diff
@@ -402,33 +405,87 @@ fn decode_otlp_logs_json(body: &[u8]) -> Result<Vec<u8>, InputError> {
...
+        let decoded = base64::engine::general_purpose::STANDARD
+            .decode(bytes)
+            .map_err(|e| InputError::Receiver(format!("invalid OTLP JSON bytesValue: {e}")))?;
+        return Ok(Some(hex::encode(&decoded)));
     }
-    String::new()
+    Ok(None)
+}
+
+/// Write an OTLP JSON AnyValue as a JSON field preserving primitive types.
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/logfwd-io/src/otlp_receiver.rs]: Fixed in `68e24be8`.

Dynamic OTLP keys now go through the same JSON escaping path as string values, and the generated `otelcol` YAML now uses safely quoted strings instead of raw `display()` output inside double-quoted YAML.

File: crates/logfwd-io/tests/it/otlp_contract_support.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    if let Some(bytes) = value.get("bytesValue").and_then(serde_json::Value::as_str) {
+        let decoded = base64::engine::general_purpose::STANDARD
+            .decode(bytes)
+            .ok()?;
+        return Some(serde_json::Value::String(hex_encode(&decoded)));
+    }
+    None
+}
+
+fn normalized_json_number(value: f64) -> Option<serde_json::Number> {
```
[the remark this answers is not in the document]
Comment (strawgate) [reply crates/logfwd-io/tests/it/otlp_contract_support.rs]: Fixed in `68e24be8`.

The numeric normalizer now uses a strict upper bound so `2^63` can no longer round into the signed-int path and silently saturate to `i64::MAX`.

**Comment by strawgate** [comment]
Self-review follow-up is in `68e24be8`.

Addressed before human review:
- malformed OTLP JSON `intValue` now errors instead of silently becoming `0`
- dynamic OTLP keys are JSON-escaped correctly
- generated `otelcol` YAML now uses safely quoted strings for paths/endpoints
- numeric oracle normalization no longer lets `2^63` round into the signed-int path

Extra targeted checks run after the fixes:
- `cargo test -p logfwd-io invalid_json_int_value_returns_error -- --nocapture`
- `cargo test -p logfwd-io json_escaping_key_chars -- --nocapture`
- `cargo test -p logfwd-io --test it`
- `cargo test -p logfwd-io --test it otelcol_ -- --ignored --nocapture`

I also tried a broader `cargo clippy -p logfwd-io -p logfwd-output --tests -- -D warnings` pass. It still reports pre-existing unrelated warnings in untouched files (`null.rs`, `generator.rs`, `segment.rs`, `lib.rs`, `checkpoint_state_machine.rs`), but there were no new clippy blockers from the follow-up changes above.

**Comment by strawgate** [comment]
Needs a rebase on master to resolve the conflict — all reviewer concerns appear to have been addressed in the latest commits. Once rebased and CI is green this should be ready.
